usb: device_next: implementation of USB device support notification

jfischer-no opened the pull request:
usb: device_next: implementation of USB device support notification

The implementation uses the system workqueue and a callback provided
and registered by the application. The application callback is called in
the context of the workqueue. Notification messages are stored in a
queue and delivered to the callback in sequence. The callback may return
error code -EBUSY or -EAGAIN, in which case the implementation retries
to deliver the message again.

We cannot call application callback directly from the USB device stack
thread because the behavior of arbitrary code provided by the
application is unpredictable, and we do not want it to be executed in
the same context where all events from the device controller are
handled.

Nor can we use the ZBUS subsystem directly. ZBUS offloads responsibility
for defined behavior to the observers and application, and does not
provide any way for the publisher to enforce defined behavior and
execution context.
ZBUS listener would be called from the USB thread context and is not
acceptable. ZBUS subscriber does not provide delivery guarantee and
cached message can be overwritten. ZBUS message subscriber has
cumbersome global configuration and buffers that are too complicated to
handle from USB configuration, and even if we would use it, defined
behavior is not possible because of how listener and subscriber are
handled.

Resolves: #51034
Resolves: #50759

jfischer-no committed:
usb: device_next: implementation of USB device support notification

The implementation uses the system workqueue and a callback provided
and registered by the application. The application callback is called in
the context of the workqueue. Notification messages are stored in a
queue and delivered to the callback in sequence.

We cannot call application callback directly from the USB device stack
thread because the behavior of arbitrary code provided by the
application is unpredictable, and we do not want it to be executed in
the same context where all events from the device controller are
handled.

Nor can we use the ZBUS subsystem directly. ZBUS offloads responsibility
for defined behavior to the observers and application, and does not
provide any way for the publisher to enforce defined behavior and
execution context.
ZBUS listener would be called from the USB thread context and is not
acceptable. ZBUS subscriber does not provide delivery guarantee and
cached message can be overwritten. ZBUS message subscriber has
cumbersome global configuration and buffers that are too complicated to
handle from USB configuration, and even if we would use it, defined
behavior is not possible because of how listener and subscriber are
handled.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: cdc_acm: use USB notification messages to detect DTR

Use USB notification messages to detect DTR and baudrate updates.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: use USB notification in CDC ACM implementation

Add message types for line coding and contol line state updates.
Add a publish message function that takes a pointer to a device
structure as payload, and use USB notification in the CDC ACM
implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on include/zephyr/usb/usbd_msg.h:
```diff
@@ -0,0 +1,124 @@
+	USBD_MSG_SUSPEND,
+	/** Bus reset detected */
+	USBD_MSG_RESET,
+	/** Reserved */
+	USBD_MSG_RESERVED0,
+	/** Reserved */
+	USBD_MSG_RESERVED1,
```
What's the point of having reserved values here? Isn't just keeping API compatibility (i.e. not renaming the entries, but reordering is fine) enough here?

tmon-nordic commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,120 @@
+
+	ctx = m_pkt->ctx;
+	err = ctx->msg_cb(&m_pkt->msg);
+	if (err == -EBUSY || err == -EAGAIN) {
+		LOG_DBG("Message callback is busy (%d)", err);
+		(void)k_work_submit(work);
+		return;
+	}
```
What's the point? This is likely to result in endless loops taking up 100% CPU time. System workqueue priority defaults to lowest cooperative priority so it is likely that whatever kept the callback busy won't have finished when system workqueue starts processing this work item again. If coop is not enabled, then it defaults to priority 0 which would be the highest priority in the system.

tmon-nordic commented on include/zephyr/usb/usbd_msg.h:
```diff
@@ -0,0 +1,124 @@
+/**
+ * @brief USB device support message types
+ *
+ * The first set of message types map to event types from the UDC driver API,
```
Silently casting the enum types and hoping that numerical value will match is bad design in my opinion. Things will break (because one enum will be updated without updating the other) during development and it'll be hard to track.

jfischer-no replied:
Right, and it does not make the code smaller.

tmon-nordic replied:
You can update the comment and remove "reserved types are not used"

tmon-nordic commented on subsys/usb/device_next/usbd_core.c:
```diff
@@ -128,33 +128,39 @@ static ALWAYS_INLINE int usbd_event_handler(struct usbd_contex *const uds_ctx,
 	case UDC_EVT_VBUS_READY:
-		LOG_WRN("VBUS detected event");
+		LOG_DBG("VBUS detected event");
+		usbd_msg_pub_simple(uds_ctx, event->type, 0);
```
Don't silently cast one enum to another! Use proper enum value here (`USBD_MSG_VBUS_READY`).

jfischer-no replied:
https://github.com/zephyrproject-rtos/zephyr/pull/68256#discussion_r1474610952

tmon-nordic commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,120 @@
+	int err;
+
+	if (ctx->msg_cb == NULL) {
+		return -ENOTSUP;
```
Propably a personal preference, but I wouldn't consider this to be an error condition. I would just `return 0;` here.

jfischer-no replied:
reworked, error can not be forwarded anyway.

tmon-nordic commented:
@rodrigopex I would appreciate your review and/or opinion on this.

rodrigopex commented:
@tmon-nordic, that is a clear reimplementation of part of what Zbus is made for. Regarding to:

> ZBUS message subscriber has
cumbersome global configuration and buffers that are too complicated to
handle from USB configuration, and even if we would use it, defined
behavior is not possible because of how listener and subscriber are
handled.

@jfischer-no, I am intrigued about the cumbersome configuration you mentioned. There are only four configurations:

- `CONFIG_ZBUS_MSG_SUBSCRIBER` to enable the feature;
- `CONFIG_ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_DYNAMIC` and `CONFIG_ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_STATIC` telling zbus the allocation of the message pool will be static or dynamic;
- At last, the `CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE` will set the size of the `net_buf_pool` as you can find in any other network or Bluetooth code. 

Maybe you see limitations I am not able to see. Please let me know if you have an idea of how to make zbus useful for your case. You may have a special channel that only enables message subscribers to observe it. Or create a way of hiding some system channels. Or a way of separating message subscriber pools: one for the user and the other for the system. It would solve your problems. 

If everyone who doesn't find a feature in zbus implements its own event manager, we will not converge in a great bus. I am here to help improve that by listening to the community's needs.

jfischer-no commented:
> @tmon-nordic, that is a clear reimplementation of part of what Zbus is made for. Regarding to:

It is not a re-implementation. Zbus (which is not a bus) does not provide this functionality and defined behavior for the publisher context (explained in the commit message).

> 
> > ZBUS message subscriber has
> > cumbersome global configuration and buffers that are too complicated to
> > handle from USB configuration, and even if we would use it, defined
> > behavior is not possible because of how listener and subscriber are
> > handled.
> 
> @jfischer-no, I am intrigued about the cumbersome configuration you mentioned. There are only five configurations:
> 
>     * `CONFIG_ZBUS_MSG_SUBSCRIBER` to enable the feature;
> 
>     * `CONFIG_ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_DYNAMIC` and `CONFIG_ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_STATIC` telling zbus the allocation of the messages will be static or dynamic;
> 
>     * `CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_STATIC_DATA_SIZE` as the message is static, we need to tell zbus what that size is;
> 
>     * At last, the `CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE` will set the size of the `net_buf_pool` as you can find in any other network or Bluetooth code.
> 
> If you are using dynamic allocation for those messages, as you are doing in your code, you will only use: `CONFIG_ZBUS_MSG_SUBSCRIBER=y` and `CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE=X`. Set the proper pool size, and that is it.

Yes, and these are some of the problems I absolutely do not want to deal with, nor do I want to force USB users to do so. These configurations are global, instantiated in zbus.c. I do not want USB to share a pool with others. I do not want to force users to enable and configure HEAP. This could be fixed by instantiating memory (net_bufs, just blocks, ...) in the specific ZBUS_CHAN_DEFINE_FOOBAR macro, but other problems remains.

> 
> Maybe you see limitations I am not able to see. Please let me know if you have an idea of how to make zbus useful for your case. You may have a special channel that only enables message subscribers to observe it. Or create a way of hiding some system channels. Or a way of separating message subscriber pools: one for the user and the other for the system. It would solve your problems.

I described "limitations" in the commit message. It cannot be fixed because of the way ZBUS is designed. The behavior of the channel depends on the type of observers and can be changed by the application, regardless of whether the channel is "hidden" or not.

> If everyone who doesn't find a feature in zbus implements its own event manager, we will not converge in a great bus. I am here to help improve that by listening to the community's needs.

It is not just a missing feature. I really see no reason to use ZBUS in Zephyr subsystems, :zero:. All the provided Zbus abstractions of kernel "data passing" can be implemented in a much cleaner and more efficient way without Zbus, but additionally with defined behavior.

carlescufi commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,143 @@
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (k_mem_slab_alloc(&usbd_msg_slab, (void **)&m_pkt, K_NO_WAIT)) {
```
not blocking here?

jfischer-no replied:
We cannot block, and should not, described in the issue.

rodrigopex commented:
> Zbus (which is not a bus) does not provide this functionality and defined behavior for the publisher context (explained in the commit message)

@jfischer-no, please specify a formal bus definition that tells us zbus is not a bus. Is that your own definition of a bus?

> It is not just a missing feature. I really see no reason to use ZBUS in Zephyr subsystems, 0️⃣. All the provided Zbus abstractions of kernel "data passing" can be implemented in a much cleaner and more efficient way without Zbus, but additionally with defined behavior.

I am curious to see the best way to do that because Zbus uses a semaphore and a memory copy to store data in the channel and deliver the messages using fifos (like you here), message queues, and callbacks. 

>  It cannot be fixed because of the way ZBUS is designed. 

Why did you not help us during the zbus submission, since you have a better way to do things? I am trying to recall the RFC and the PR adding Zbus; your name is not there. 

I am confident we can solve this pool-sharing issue and other zbus gaps.

jfischer-no committed:
samples: usb: add message callback registration to the common code

This allows samples to receive earlier notification immediately after
initialization.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,127 @@
+
+void usbd_msg_pub_simple(struct usbd_contex *const ctx,
+			 const enum usbd_msg_type type, const int status)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (ctx->msg_cb == NULL) {
+		return;
+	}
+
+	m_pkt = usbd_alloc_msg(ctx);
+	if (m_pkt == NULL) {
+		__ASSERT(false, "Failed to allocate message memory");
+		return;
+	}
+
+	m_pkt->msg.type = type;
+	m_pkt->msg.status = status;
+	submit_new_work(m_pkt);
+}
+
+void usbd_msg_pub_device(struct usbd_contex *const ctx,
+			 const enum usbd_msg_type type, const struct device *const dev)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (ctx->msg_cb == NULL) {
+		return;
+	}
+
+	m_pkt = usbd_alloc_msg(ctx);
+	if (m_pkt == NULL) {
+		__ASSERT(false, "Failed to allocate message memory");
+		return;
+	}
+
+	m_pkt->msg.type = type;
+	m_pkt->msg.dev = dev;
+	submit_new_work(m_pkt);
+}
```
?

```c
static inline void usbd_msg_pub(struct usbd_contex *const ctx,
				const struct usbd_msg msg)
{
	struct usbd_msg_pkt *m_pkt;

	if (k_mem_slab_alloc(&usbd_msg_slab, (void **)&m_pkt, K_NO_WAIT)) {
		LOG_ERR("Failed to allocate message memory");
		__ASSERT(false, "Failed to allocate message memory");
		return;
	}

	m_pkt->ctx = ctx;
	memcpy(&m_pkt->msg, &msg, sizeof(msg));

	k_fifo_put(&msg_queue, m_pkt);
	if (k_work_submit(&msg_work) < 0) {
		__ASSERT(false, "Failed to submit work");
	}
}
```

This makes the usbd_msg memory footprint a little larger.

carlescufi replied:
In that case let's leave it as it is

tmon-nordic commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,127 @@
+
+	m_pkt = usbd_alloc_msg(ctx);
+	if (m_pkt == NULL) {
+		__ASSERT(false, "Failed to allocate message memory");
```
This doesn't quality for an ASSERT, because this is somewhat expected that the slab can run empty. Asserts are for things that indicate software bug.

jfischer-no commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -0,0 +1,127 @@
+
+	m_pkt = usbd_alloc_msg(ctx);
+	if (m_pkt == NULL) {
+		__ASSERT(false, "Failed to allocate message memory");
```
There is no other channel to report this, the alternative is LOG_ERR or k_panic() or to block in alloc(). @tmon-nordic You preferences?

desowin replied:
LOG_ERR doesn't make sense here because it will flood the log (if we had equivalent to Linux log once methods then maybe...). k_panic is similar to assert here and therefore not proper. Blocking in alloc is better than other alternatives. There's also possibility to simply not log it at all and not block.

tmon-nordic replied:
LOG_ERR doesn't make sense here because it will flood the log (if we had equivalent to Linux log once methods then maybe...). k_panic is similar to assert here and therefore not proper. Blocking in alloc is better than other alternatives. There's also possibility to simply not log it at all and not block.

jfischer-no replied:
> There's also possibility to simply not log it at all and not block.

Would it not be equivalent to assert abuse? assert is also used in the tree for troubleshooting.

desowin replied:
Here only tracing is appropriate. Assert is used for troubleshooting because it is designed to expose incorrect code, e.g. calling function with invalid parameters and ao on. Abusing asserts for normal runtime behavior is wrong.

tmon-nordic replied:
Here only tracing is appropriate. Assert is used for troubleshooting because it is designed to expose incorrect code, e.g. calling function with invalid parameters and ao on. Abusing asserts for normal runtime behavior is wrong.

jfischer-no replied:
@tmon-nordic Please recheck.

rodrigopex commented:
@jfischer-no, I prefer to comment only on the facts.

> but from the publisher perspective you never know what happens, there could be listeners or subscribers, some global configurations enabled or disabled, global setting for all channels....

Listeners and subscribers consume the data differently but receive the message when it is well implemented. We added global configurations and variations to solve community needs.

> Btw, ZBUS is many channels with 1 -> n communication, not 1 <-> n, not m <-> n.

One-to-one, one-to-many, and many-to-many are possible with ZBus. When we say many-to-many, we are not implying many-toAndFrom-many. A channel-based bus forces us to think in different channels for different contexts. Suppose that worked in many-toAndFrom-many (m <-> n). You could use only one channel for everything, which is the idea of sockets but not a bus.

jfischer-no commented:
> @jfischer-no, I prefer to comment only on the facts.
> 
> > but from the publisher perspective you never know what happens, there could be listeners or subscribers, some global configurations enabled or disabled, global setting for all channels....
> 
> Listeners and subscribers consume the data differently but receive the message when it is well implemented. We added global configurations and variations to solve community needs.
> 
> > Btw, ZBUS is many channels with 1 -> n communication, not 1 <-> n, not m <-> n.
> 
> One-to-one, one-to-many, and many-to-many are possible with ZBus. When we say many-to-many, we are not implying many-toAndFrom-many. A channel-based bus forces us to think in different channels for different contexts. Suppose that worked in many-toAndFrom-many (m <-> n). You could use only one channel for everything, which is the idea of sockets but not a bus.

That is your opinion.

jfischer-no commented:
CI fails because of https://github.com/zephyrproject-rtos/zephyr/issues/68479

tmon-nordic commented on include/zephyr/usb/usbd_msg.h:
```diff
@@ -57,6 +61,8 @@ static const char *const usbd_msg_type_list[] = {
 	"Bus reset",
 	"Controller error",
 	"Stack error",
+	"New CDC ACM line coding",
+	"New CDC ACM control line state",
```
Nit: this is not necessarily "New" as the line coding or control line state might be exactly the same as before (the notifications are for every SET_LINE_CODING and SET_CONTROL_LINE_STATE).

jfischer-no replied:
Yes, I am bad at producing human-readable text.

tmon-nordic reviewed (dismissed):
Just one nit, but that can be revised later. Not sure if system workqueue is better than synchronous notifications, but we'll have to test it in the wild.

tmon-nordic reviewed (approved):
Approve, but my gut feeling is that the asynchronous notifications will cause troubles and there'll be push towards synchronous callbacks. (in USB thread context).

lopsided98 commented:
#70417 added a new call to `sample_usbd_init_device()` right before this PR was merged and needs to be updated.

lopsided98 commented:
Yes, that was the fix.
